### ISSUE TITLE
Support direct attachment of volumes to the VM

### ISSUE THREAD

**Issue by fabiand** [kubevirt/kubevirt#952]
Today, volumes are mounted by the kublet, then passed to the VM for consumption.
This is well integrated, but we pay a price.

Another option is to o connect volumes directly to the qemu process (for whatever protocol qemu supports), this would avoid the intermediate layers, but has other drawbacks.

Main benefits are: Less user-space and kernel space dependencies, more rpedictive behavior on VM shutdown, different performance charactistics.

**Comment by fabiand** [comment]
Two discussion points were:

a. direct attached volumes will behave differently than volumes (because they use a different tech stack) - This needs to be somehow implicitly visible from the API
b. Naively seen, qemu will need to access the storage targets. However, we can not leak the PV details into the VM Pod, as PV connection details are considered to be pirvate

WRT a.: Consider to not use `volume:` for directly attached storage, instead only keep it in the `disk:` definition
WRT b.: Consider to have a sidecar pod which performs the actual PV lookup and storage connection, and provides an NBD socket to the launcher pod, to use it with qemu. Here the performance drawback has to be understood.

**Comment by vitalif** [comment]
Why is this closed?) This is a 100% MUST)

**Comment by vitalif** [comment]
Please reopen :-)

**Comment by xpivarc** [comment]
@vitalif Can you elaborate on why we need this?

**Comment by vitalif** [comment]
> @vitalif Can you elaborate on why we need this?

Kernel mounting is an overhead and may in theory even be impossible for some storages... For example, I don't have a kernel driver for my project Vitastor :) I provide support via NBD, but it's much slower than native driver. https://yourcmc.ru/git/vitalif/vitastor/src/branch/master/docs/performance/comparison1.en.md#vitastor-0-4-0-native - also NBD has a problem with timeout - if you don't specify a timeout then if the mount process fails it becomes impossible to remove device from system without rebooting; and if specify a timeout then the device autostops when an operation is suspended for more than timeout seconds which is also an undesirable effect...

"Unkillable" devices may also appear in case of Ceph RBD when there is a process in D state stuck reading or writing the device, and you'll also have to reboot.

QEMU userspace drivers don't have this problem... Also QEMU probably provides additional features for userspace block drivers... Most popular qemu-based hypervisors use userspace QEMU drivers, so I think kubevirt must also have this feature.

**Comment by vitalif** [comment]
I have my own CSI based on NBD and I don't expose any sockets - NBD in my case works as "FUSE for block devices", it has such mode. So I don't think I need any extra CSI :)
And NBD is an overhead in any case. Data goes from qemu to the nbd server, then from nbd to the cluster. Without NBD data goes directly to the cluster. I don't think it's a good idea to sacrifice 50% of random i/o performance and extra CPU usage.
By the way, what's the launcher and what credentials are you talking about? Isn't the thing that runs QEMU some kind of a system service, restricted from direct user access?

**Comment by alicefr** [comment]
@fabiand it isn't totally clear to me the advantage that this approach will bring. You need in any case to mount the socket in the pod and make it available to QEMU

As @vitalif already pointed out you can also use TCP instead of UNIX socket and if the NBD endpoint is reachable from the VM pod then you can simply access it. Maybe using vhost user protocol could be a more performant solution, but again I don't completely understand why we want to access the storage indirectly.

Generally, I'd encourage a CSI plugin if we want also to implement the storage provisioning, but here I don't think it is the case., right?

**Comment by vitalif** [comment]
> You can also use TCP instead of UNIX socket

@alicefr Just to be more clear, I'll repeat the main point. I don't need NBD. I need QEMU to load `block-vitastor.so` and access my cluster directly. It can't be wrapped in another network protocol because I already have **my own** protocol. And it's distributed, client connects to all storage nodes, there is no single "gateway".

And I don't need vhost-user either. I need native QEMU block driver. block-rbd.so block-vitastor.so block-nfs.so and so on. I want kubevirt to support these!

**Comment by alicefr** [comment]
@vitalif thanks, sorry it wasn't clear at the beginning. Let me rephrase, what you need is a special driver in order to pass your storage and currently, there isn't such an option in KubeVirt right? How does QEMU access the storage? More precisely how do you want to expose the storage to the container where the VM is running?

**Comment by vitalif** [comment]
QEMU accesses my storage over the network, same as Ceph (block-rbd.so). Then it uses a userspace library and, optionally, credentials to connect to the cluster and send i/o requests directly from the qemu process. To use that all you need to do is pass a correct block device string to QEMU, like "-drive format=raw,file=rbd:testimg". Then in case of Vitastor it doesn't use authentication (it's supposed to just block the network access from inside VMs to the cluster somehow), in case of Ceph typically a single user is used for all VM images and its credentials (username, key) are read from /etc/ceph/ by the library used by qemu. And kubevirt can't do that, it requires to expose the image via kernel, partly because CSI doesn't know anything about QEMU block drivers

**Comment by alicefr** [comment]
Yes, we don't support all the QEMU block drivers in KubeVirt. The storage is passed with virtio blk and as [driver QEMU](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/converter/converter.go#L202-L207) also for ceph we don't use the rbd protocol. 
As a quick try you could modify the libvirt XML and pass the options you need using a [sidecar](https://github.com/kubevirt/kubevirt/tree/main/cmd/example-hook-sidecar) (here an [example]()) if the network is accessible from inside the pod.

Afterward, there have been in the past some ideas of using [qemu-storage-daemon](https://qemu.readthedocs.io/en/latest/tools/qemu-storage-daemon.html) in order to separate the QEMU storage functionality. We could generalize the access to the storage by using the qsd and exporting the storage to qemu through a socket (see the `--export` options). This approach adds a level of indirection but it simplifies what KubeVirt needs to support in terms of storage driver types and managment. Does this sound a long term alternative to you?

**Comment by vitalif** [comment]
There are 2 main benefits of direct access:
1) increased throughput due to data copy avoidance
2) reduced latency due to less IPC when i/o is done directly from the QEMU main event loop. Improvement is around ~0.04ms per 4 kb I/O which is not a lot, but still noticeable for SSDs

If QSD can provide these 2 benefits too then I think it's okay(ish), if not then it's not much better than NBD for me.

Also after reading the documentation of QSD I'm not sure. Does vhost-user-blk use a UNIX socket or shared memory for data? If it uses a UNIX socket then it's no better than NBD, if it uses shared memory it can in theory at least solve (1), but not (2).

There are also 2 smaller issues:
3) QSD can provide support for hypothetical storages that have a QEMU driver BUT don't have neither kernel driver nor iSCSI nor NBD gateway nor CSI driver nor ...anything else :). Though it doesn't sound like something existing in reality, I doubt if such storages exist at all. 
4) QSD can in theory solve my NBD timeout problem. However I suppose that considering this issue alone it can also be solved just by exposing an NBD UNIX socket and using it from kubevirt directly using QEMU user-space nbd driver instead of the kernel...

So I think these 2 additional points are rather insignificant.

I'll try libvirt, thanks. By the way, libvirt also doesn't have a good plugin architecture, it has a hardcoded block device driver list and I also patch it for Vitastor support :). The best plugin architecture is in Proxmox. The block device is presented by a single Perl file (.pm) there and libvirt is not used at all :)

**Comment by vitalif** [comment]
>AFAIU, the qsd uses the shared memory over the UNIX socket using vhost-user protocol. It should not have any performance penalty but I'm not sure if we already have measured this.

How does it work over the socket? I tried virtiofsd once and it required access to the whole QEMU address space. Anyway single thread is better in terms of latency than 2 threads - QEMU and external process - having to communicate through any kind of IPC. I.e. if it uses shared memory then memory copying is probably not an issue but latency still suffers.

>I'm not sure about the other points you mentioned. Could you please give more details about the reason you need them?

It was just theory :) What I want is direct block driver access. `-drive file=vitastor:etcd_host=...:image=testimg`. It's the simplest way) other options are more complex and have non-obvious advantages.

>Just curious do you have any plans to integrate your driver in QEMU upstream?

I should probably at least try to offer it to them, yeah. But I'm not sure if they'll easily take it given that libvitastor-client isn't included in any major distros :)

**Comment by taljeth** [comment]
> Also after reading the documentation of QSD I'm not sure. Does vhost-user-blk use a UNIX socket or shared memory for data? If it uses a UNIX socket then it's no better than NBD, if it uses shared memory it can in theory at least solve (1), but not (2).

It uses the Unix socket as a control channel to set up everything, but I/O is done with shared memory. Essentially instead of the QEMU binary processing the virtio-blk ring, you do the processing in the QSD process. The advantage in your scenario is that you have full control over what that QSD process does, so you get access to all of QEMU's block backend functionality, including your custom driver.

Note that this only really works when you use virtio-blk in the guest, though. While we are working on a block driver that would allow attaching a vhost-user-blk export from QSD to any device in QEMU, this wouldn't be able to use shared memory any more, so without virtio-blk it wouldn't be very different from NBD any more.

> There are also 2 smaller issues: 3) QSD can provide support for hypothetical storages that have a QEMU driver BUT don't have neither kernel driver nor iSCSI nor NBD gateway nor CSI driver nor ...anything else :). Though it doesn't sound like something existing in reality. I doubt if such storages exist at all.

Or for storage where other options like an NBD gateway exist, but aren't the best solution. In other words, your exact case. :)

> How does it work over the socket? I tried virtiofsd once and it required access to the whole QEMU address space.

It requires access not to the whole QEMU address space, but to the guest RAM. You (or Kubevirt really) need to configure a memfd memory backend so that we have a file descriptor that can be shared over the Unix socket.

> Anyway single thread is better in terms of latency than 2 threads - QEMU and external process - having to communicate through any kind of IPC. I.e. if it uses shared memory then memory copying is probably not an issue but latency still suffers.

There is no inherent reason why notifying the QEMU process of new requests (or the guest of request completions from QEMU) would be faster than notifying the QSD process.

What could well be is that the vhost-user-blk export hasn't received as much optimisation work yet as the virtio-blk device implementation in QEMU, so while the performance should in theory be able to be the same, in practice I would expect some differences.

**Comment by vitalif** [comment]
> There is no inherent reason why notifying the QEMU process of new requests (or the guest of request completions from QEMU) would be faster than notifying the QSD process.

Maybe... we need some benchmarks to decide :) if the guest notifies QSD directly without switching to QEMU event loop first, then it may be the same, yeah. It may even be slightly better because I/O will be done in a separate process and won't share the CPU core with the rest of QEMU event loop. So yeah, in this case it's also an option :)

**Comment by vitalif** [comment]
> Note that this only really works when you use virtio-blk in the guest, though

It's fine, other drivers are worse anyway :)

**Comment by vitalif** [comment]
So, okay, in this case if it looks better to you to implement qemu block driver support via QSD+vhost-user it looks fine to me, too :)

**Comment by stefanha** [comment]
Generalizing a bit, I think KubeVirt should support vhost-user-blk. This allows another process on the same host to handle virtio-blk I/O without adding overhead. This architecture also keeps storage access credentials out of the QEMU process so a compromised QEMU process cannot steal the credentials.

Once KubeVirt supports vhost-user-blk UNIX domain sockets then @vitalif can build a qemu-storage-daemon (QSD) container image with the vitastore QEMU block driver and it can be used with KubeVirt. Perhaps the QEMU project can release an official QSD container image in the future too.

In the long term it may be possible to introduce a new volume type in k8s for vhost-user-blk. It would be an alternative to file and block volumes for applications that speak vhost-user-blk natively, like KubeVirt/QEMU. Then the vitastore CSI plugin could advertise vhost-user-blk support and avoid the NBD performance penalty.

**Comment by alicefr** [comment]
>It requires access not to the whole QEMU address space, but to the guest RAM. You (or Kubevirt really) need to configure a memfd memory backend so that we have a file descriptor that can be shared over the Unix socket.

Memfd is already the default memory backend in KubeVirt, so nothing to do on this front :)

>Once KubeVirt supports vhost-user-blk UNIX domain sockets then @vitalif can build a qemu-storage-daemon (QSD) container image with the vitastore QEMU block driver and it can be used with KubeVirt. At that point CSI is being bypassed though.

For this we could define a new couple of [volume](https://kubevirt.io/user-guide/virtual_machines/disks_and_volumes/#volumes) and [disk](https://kubevirt.io/user-guide/virtual_machines/disks_and_volumes/#disks) types. Then, we need a privileged component that perform the mounting inside the VM pod. This could be done by KubeVirt if it knows where to find the socket or by another process like it happened for CSI. 
All this is possible, but we need a [design proposal](https://github.com/kubevirt/community/blob/main/design-proposals/README.md) and a couple of use cases to justify the change in KubeVirt.

>In the long term it may be possible to introduce a new volume type in k8s for vhost-user-blk. It would be an alternative to file and block volumes for applications that speak vhost-user-blk natively, like KubeVirt/QEMU. Then the vitastore CSI plugin could advertise vhost-user-blk support and avoid the NBD performance penalty. That would allow the vitastore CSI plugin to be used again instead of a KubeVirt-specific vhost-user-blk connection.

I'm not sure if this will be possible. I think in order to be accepted in CSI, it would need to be useful for standard containers or adopted by other projects (maybe kata containers). So, I'd suggest to focus for now to have a good integration with KubeVirt.

**Comment by stefanha** [comment]
> I'm not sure if this will be possible. I think in order to be accepted in CSI, it would need to be useful for standard containers or adopted by other projects (maybe kata containers). So, I'd suggest to focus for now to have a good integration with KubeVirt.

Yes, Kata Containers could connect to vhost-user-blk volumes. The device would appear as a block volume to the container running inside the sandbox VM.

You know KubeVirt, CSI, and k8s better than I do. If you think a good step is to define a KubeVirt volume type for vhost-user-blk then that sounds good.

**Comment by alicefr** [comment]
We will probably need some thinking about vhost-user integration. A list of possible use-case will be very helpful to understand how this will be used.
@vitalif could you please let us know how it goes with the QSD and your custom driver.

**Comment by alicefr** [comment]
@vitalif one additional suggestion. 
From your description, you don't need to mount anything particular to the VM pod, and the storage should be reachable from the network. You could build the QSD with your driver and inject it into the virt-launcher pod. This can be achieved with a webhook (like istio is injecting envoy as sidecar). Once you have an additional container with your changes, you can share a directory between the virt-launcher container and the QSD container, so you can make available the vhost-user socket to libvirt. Unfortunately, you still need a second sidecar to modify the XML in order to pass the socket but this should be a relative simple and viable solution for your setup. You don't need to path QEMU and libvirt and you can have your own container images.
